### PR TITLE
Opencv4 tests and blas provider

### LINKED ISSUE #196653: python3Packages.opencv4 depends on multiple versions of OpenBLAS, segfaults when used with scipy
### Describe the bug
I have a Python script that uses opencv and scikitimage, and it began segfaulting after updating nixpkgs from 9921c621ed25f7a2c64a7e3a95101e9ef7e8a4e6 to d9ef9f36454393adaee6674e978c0caee82f979a.

### Steps To Reproduce

From a git checkout of nixpkgs at d9ef9f36454393adaee6674e978c0caee82f979a, the minimal reproducer is:

```console
~/nixpkgs$ nix-shell -E 'with import ./. {}; (python3.withPackages(ps:[ ps.opencv4 ps.scipy ])).env' 

[nix-shell:~/nixpkgs]$ python3
Python 3.10.7 (main, Sep  5 2022, 13:12:31) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import cv2
>>> import scipy
>>> scipy.linalg.inv([[1, 0, 0],[0, 1, 0],[0, 0, 1]])
Segmentation fault (core dumped)
```

(scikitimage depends on scipy, and the segfault is in a call to `scipy.linalg.inv` during its initialization)

The order of the imports matters; when importing `scipy` first, it doesn't segfault.

### Additional context

GDB backtrace of the segfault:

```
(gdb) bt
#0  0x00007ffff2e7a4e7 in idamax_k_ZEN () from /nix/store/qb2h3qwksr6f3zjs78dv5gray8rjsgzz-openblas-0.3.21/lib/libopenblas.so.0
#1  0x00007ffff21d6926 in dgetf2_k () from /nix/store/qb2h3qwksr6f3zjs78dv5gray8rjsgzz-openblas-0.3.21/lib/libopenblas.so.0
#2  0x00007ffff21d4643 in dgetrf_parallel () from /nix/store/qb2h3qwksr6f3zjs78dv5gray8rjsgzz-openblas-0.3.21/lib/libopenblas.so.0
#3  0x00007ffff21d4643 in dgetrf_parallel () from /nix/store/qb2h3qwksr6f3zjs78dv5gray8rjsgzz-openblas-0.3.21/lib/libopenblas.so.0
#4  0x00007ffff21d4643 in dgetrf_parallel () from /nix/store/qb2h3qwksr6f3zjs78dv5gray8rjsgzz-openblas-0.3.21/lib/libopenblas.so.0
#5  0x00007ffff21d4643 in dgetrf_parallel () from /nix/store/qb2h3qwksr6f3zjs78dv5gray8rjsgzz-openblas-0.3.21/lib/libopenblas.so.0
#6  0x00007ffff1fb9975 in dgetrf_ () from /nix/store/qb2h3qwksr6f3zjs78dv5gray8rjsgzz-openblas-0.3.21/lib/libopenblas.so.0
#7  0x00007fffa9812970 in f2py_rout.flapack_dgetrf () from /nix/store/hz2z9q2zfzf4c33il7n5x99zrp7qdfmf-python3-3.10.7-env/lib/python3.10/site-packages/scipy/linalg/_flapack.cpython-310-x86_64-linux-gnu.so
#8  0x00007ffff7c95b24 in _PyObject_MakeTpCall () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#9  0x00007ffff7c48ce9 in _PyEval_EvalFrameDefault () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#10 0x00007ffff7dc3a8f in _PyEval_Vector () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#11 0x00007ffff7c484ce in _PyEval_EvalFrameDefault () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#12 0x00007ffff7dc3a8f in _PyEval_Vector () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#13 0x00007ffff7dc40f8 in PyEval_EvalCode () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#14 0x00007ffff7e487ed in run_mod () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#15 0x00007ffff7e54af0 in PyRun_InteractiveOneObjectEx () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#16 0x00007ffff7e54e66 in _PyRun_InteractiveLoopObject () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#17 0x00007ffff7e55cbe in PyRun_AnyFileExFlags () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#18 0x00007ffff7e59f49 in Py_RunMain () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#19 0x00007ffff7e5a185 in Py_BytesMain () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#20 0x00007ffff78bf24e in __libc_start_call_main () from /nix/store/c6f52mvbv0d8rd3rlslsvy7v4g3pmm7p-glibc-2.35-163/lib/libc.so.6
#21 0x00007ffff78bf309 in __libc_start_main_impl () from /nix/store/c6f52mvbv0d8rd3rlslsvy7v4g3pmm7p-glibc-2.35-163/lib/libc.so.6
#22 0x0000000000401075 in _start ()
```

I then set a breakpoint on `idamax_k_ZEN` without importing `cv2`, and noticed that the nix store path of the `.so` containing this function differs when the crash does not occur:

```
(gdb) bt
#0  0x00007fffce5da400 in idamax_k_ZEN () from /nix/store/fgmgppypxs3mfn1qp1sibi9z7c5cm9wv-openblas-0.3.21/lib/libopenblas.so.0
#1  0x00007fffcd933a60 in dgetf2_k () from /nix/store/fgmgppypxs3mfn1qp1sibi9z7c5cm9wv-openblas-0.3.21/lib/libopenblas.so.0
#2  0x00007fffcd71169e in dgetrf_ () from /nix/store/fgmgppypxs3mfn1qp1sibi9z7c5cm9wv-openblas-0.3.21/lib/libopenblas.so.0
#3  0x00007fffad40f970 in f2py_rout.flapack_dgetrf () from /nix/store/hz2z9q2zfzf4c33il7n5x99zrp7qdfmf-python3-3.10.7-env/lib/python3.10/site-packages/scipy/linalg/_flapack.cpython-310-x86_64-linux-gnu.so
#4  0x00007ffff7c95b24 in _PyObject_MakeTpCall () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#5  0x00007ffff7c48ce9 in _PyEval_EvalFrameDefault () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#6  0x00007ffff7dc3a8f in _PyEval_Vector () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#7  0x00007ffff7c484ce in _PyEval_EvalFrameDefault () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#8  0x00007ffff7dc3a8f in _PyEval_Vector () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#9  0x00007ffff7dc40f8 in PyEval_EvalCode () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#10 0x00007ffff7e487ed in run_mod () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#11 0x00007ffff7e54af0 in PyRun_InteractiveOneObjectEx () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#12 0x00007ffff7e54e66 in _PyRun_InteractiveLoopObject () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#13 0x00007ffff7e55cbe in PyRun_AnyFileExFlags () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#14 0x00007ffff7e59f49 in Py_RunMain () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#15 0x00007ffff7e5a185 in Py_BytesMain () from /nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/libpython3.10.so.1.0
#16 0x00007ffff78bf24e in __libc_start_call_main () from /nix/store/c6f52mvbv0d8rd3rlslsvy7v4g3pmm7p-glibc-2.35-163/lib/libc.so.6
#17 0x00007ffff78bf309 in __libc_start_main_impl () from /nix/store/c6f52mvbv0d8rd3rlslsvy7v4g3pmm7p-glibc-2.35-163/lib/libc.so.6
```

These store paths correspond to `openblas` and `openblasCompat` respectively:

```console
~/nixpkgs$ nix-build -A 'openblas' --no-out-link
/nix/store/qb2h3qwksr6f3zjs78dv5gray8rjsgzz-openblas-0.3.21
~/nixpkgs$ nix-build -A 'openblasCompat' --no-out-link
/nix/store/fgmgppypxs3mfn1qp1sibi9z7c5cm9wv-openblas-0.3.21
```

`opencv4` depends on both of these, while `scipy` depends only on `openblasCompat`.

```console
~/nixpkgs$ nix path-info -rSsh $(nix-build -A 'python3Packages.opencv4' --no-out-link) | grep openblas
/nix/store/fgmgppypxs3mfn1qp1sibi9z7c5cm9wv-openblas-0.3.21                       27.1M   68.3M
/nix/store/qb2h3qwksr6f3zjs78dv5gray8rjsgzz-openblas-0.3.21                       27.2M   68.4M

~/nixpkgs$ nix path-info -rSsh $(nix-build -A 'python3Packages.scipy' --no-out-link) | grep openblas
/nix/store/fgmgppypxs3mfn1qp1sibi9z7c5cm9wv-openblas-0.3.21                       27.1M   68.3M
```

So, it seems what is happening here is that `scipy` requires `openblasCompat`, but when importing `cv2` first, we get the regular `openblas`, and it crashes.

### Notify maintainers
@ttuegel @mdaiter @basvandijk

### Metadata

```console
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.0-1020-aws, Ubuntu, 20.04.4 LTS (Focal Fossa), nobuild`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.10.3`
 - channels(ubuntu): `"nixpkgs-22.11pre400930.6c6409e965a"`
 - nixpkgs: `/home/ubuntu/.nix-defexpr/channels/nixpkgs`
```

### PR DESCRIPTION
**Pull request by gbtb**
###### Description of changes
This PR is derived from my attempt to fix #196653 . I've figured out that by default blas provider uses openblas under the hood, so the change was trivial albeit a little ugly because of  `blas.provider.pname` check. Everything seemed fine, but I decided to throw in some unit tests just to be sure I didn't break anything. Initially for testing I overrided blas.provider with mkl, ran tests with it and managed to enable 80% of the opencv's tests. After that I switched blas.provider to openblas, ran tests and ... they broke, hung indefinitely, to be precise :expressionless: . I tried to set --test_threads=1 , turned off several test groups from `testNames` but different tests hung anyway. Ok, I've tried to ran tests without my blas.provider changes - result was the same :exploding_head: . Then I saw `singleThreaded` in openblas, and soon found [link to the openblas docs](https://github.com/xianyi/OpenBLAS/wiki/Faq/4bded95e8dc8aadc70ce65267d1093ca7bdefc4c#multi-threaded) and matched it with `-DWITH_OPENMP=ON` flag we use here. I've made experiments with this two flags, and results are quite logical - you can either use singleThreaded openblas, or build opencv with WITH_OPENMP=OFF, and then tests pass. So, I've figured out we have 3 options:

1. Delete tests and just use blas.provider. Nobody had any complaints so far, right? :see_no_evil:
2. Use singleThreaded openblas
3. build opencv with WITH_OPENMP=OFF

I've ruled out 1st option, because concurrency bugs could be really hard to spot and reproduce, but when they finally hit you it could cause you a lot of headache. Between 2nd and 3rd option I've intuitively inclined toward 2nd option, because I think that linear algebra calculations should be only a small part of what opencv does. Moreover, it seems that blas support was initially added to fix perfomance bottlenecks, but later optimizations were applied and [the need has vanished](https://github.com/opencv/opencv/wiki/ChangeLog#version33). But anyway, I decided to enable perf tests as well and compare both variants - in my unscientific benchmark (6C 12T i5-11400F + 32Gb)  difference of total time between two versions was negligible. So, I decided that 2nd option is "good enough" (however, documentation on multithreading support of opencv and openblas contain hints to other possible approaches with pluggable MT backends, if anyone interested to dig this rabbit hole even more)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by gbtb** [comment]
Seems like xvfb-run is unsupported on darwin, and one tests fails on arm :disappointed:

**Comment by gbtb** [comment]
Ok, x86_64-darwin first had a timeout while building openblas, 2nd try it had a timeout during run of opencv tests. I suspect something is not right with build agent itself and I would disable tests for this platform, because I don't have a machine to test them myself.

File: pkgs/development/libraries/opencv/4.x.nix [hunk 1]
```diff
@@ -208,6 +216,43 @@ let
...
+    "core"
+    "features2d"
+    "imgcodecs"
+    "imgproc"
+    "objdetect"
+    "photo"
+    "stitching"
+    "video"
+  ] ++ lib.optionals (!stdenv.isAarch64 && enableGStreamer) [ "gapi" ];
+  testRunner = if stdenv.isDarwin then "" else "${xvfb-run}/bin/xvfb-run -a ";
```
Comment (Artturin) [inline pkgs/development/libraries/opencv/4.x.nix:251]:
```suggestion
  testRunner = if stdenv.isDarwin then "" else "${lib.getExe xvfb-run} -a ";
```

**Comment by Artturin** [comment]
should target staging https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging

**Comment by gbtb** [comment]
aarch64-linux failed because of cargo:
```
error: builder for '/nix/store/ap2shazwiv9lzc6vfk5shbwiqpkab7cp-rustc-1.66.0.drv' failed with exit code 2;
       last 10 log lines:
       >           /nix/store/4xvwn9mz1zkw07nngc7bjzxxb22z5f3i-binutils-2.39/bin/ld: /build/rustc-1.66.0-src/build/aarch64-unknown-linux-gnu/stage0-rustc/aarch64-unknown-linux-gnu/release/deps/librustc_driver-2b892224648aedac.so: undefined reference to `__aarch64_cas1_acq'
       >           collect2: error: ld returned 1 exit status
       >           
       >   = help: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
       >   = note: use the `-l` flag to specify native libraries to link
       >   = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
       >
       > error: could not compile `rustc-main` due to previous error
       > Build completed unsuccessfully in 0:04:25
       > make: *** [Makefile:12: all] Error 1
       For full logs, run 'nix log /nix/store/ap2shazwiv9lzc6vfk5shbwiqpkab7cp-rustc-1.66.0.drv'.
```

**Comment by Artturin** [comment]
as far as i can tell the darwin failures are unrelated

**Comment by gbtb** [comment]
Oookey, I've refactored out tests into a separate `runCommand` derivation. Main derivation now have two outputs - out and package_tests with binaries and some data for tests. I've also remove qimgv from passthru.tests on darwin since it's not available for this platform (though it's still failing because of clang build timeout). I think now it's ready for review. @Artturin  Can you look again, pls ?

File: pkgs/development/libraries/opencv/4.x.nix [hunk 2]
```diff
@@ -368,16 +386,23 @@ stdenv.mkDerivation {
...
+        inherit opencv4;
+        };
+      };
   } // lib.optionalAttrs enablePython { pythonPath = [ ]; };
 
   meta = with lib; {
     description = "Open Computer Vision Library with more than 500 algorithms";
     homepage = "https://opencv.org/";
     license = with licenses; if enableUnfree then unfree else bsd3;
-    maintainers = with maintainers; [ mdaiter basvandijk ];
```
Comment (Artturin) [inline pkgs/development/libraries/opencv/4.x.nix:380]: why remove mdaiter?

Comment (gbtb) [reply pkgs/development/libraries/opencv/4.x.nix:380]: He is not active anymore. In the nixpkgs and elsewhere over github, it seems. Last PR in nixpkgs in [2017](https://github.com/NixOS/nixpkgs/pulls?q=is%3Apr+author%3Amdaiter+is%3Aclosed), no activity in his github profile over the last year. Personally, I think we better to not delude ourselves and remove maintainers after some period (e.g. a year) of inactivity.

File: pkgs/development/libraries/opencv/4.x.nix [hunk 3]
```diff
@@ -93,6 +95,13 @@ let
     sha256 = "sha256-meya0J3RdOIeMM46e/6IOVwrKn3t/c0rhwP2WQaybkE=";
   };
 
+  testDataSrc = fetchFromGitHub {
+    owner = "opencv";
+    repo = "opencv_extra";
+    rev = version;
+    sha256 = "sha256-6hAdJdaUgtRGQanQKuY/q6fcXWXFZ3K/oLbGxvksry0=";
+  };
```
Comment (Artturin) [inline pkgs/development/libraries/opencv/4.x.nix:103]: moved this here from tests.nix so people won't forget to update the sha when updating the version

File: pkgs/development/libraries/opencv/4.x.nix (hunk 2 above)
Comment (Artturin) [reply pkgs/development/libraries/opencv/4.x.nix:380]: added a comment about removing him to the commit msg so people don't think it was an accident

**Comment by gbtb** [comment]
Hi, @Artturin :wave: . Is there something else left preventing this PR from being merged ? I'd like to make a new release of my Stable Diffusion flake soon, hopefully with this fixes present in nixpkgs :wink:

**Pull request merged**